Tools/uorb_graph/create.py

```diff
@@ -68,189 +70,216 @@ def __init__(self, is_publication, topic_blacklist, orb_pub_sub_regexes, special
         self._topic_colors = {} # key = topic, value = color (html string)
 
-    # handle special cases
-    # format: list of tuples with 4 entries:
-    # - module name to match (module MAIN)
-    # - regex for file name(s) to match within the module (matched against the full path)
-    # - regex to extract the topic name: the match must be ORB_ID(<topic_name>
-    #   Note: whitespace is removed from source code, so it does not need to be
-    #   accounted for in the regex.
-    #   If this is None, it will just be ignored
-    # - regex to ignore matches in the form orb_[subscribe|advertise](<match>
-    #   (the expectation is that the previous matching ORB_ID() will be passed
-    #   to this, so that we can ignore it)
-        special_cases_sub = [
-    ('listener', r'.*', None, r'^(id)$'),
-    ('logger', r'.*', None, r'^(topic|sub\.metadata|_polling_topic_meta)$'),
-
-    ]
-        special_cases_sub = [(a, re.compile(b), re.compile(c) if c is not None else None, re.compile(d))
-                                   for a,b,c,d in special_cases_sub]
-
-        self._subscriptions = PubSub(False, topic_blacklist,
-                [r"\borb_subscribe(_multi|)\b\(([^,)]+)"],
-                special_cases_sub)
-
-
-        special_cases_pub = [
-    ('replay', r'Replay\.cpp$', None, r'^sub\.orb_meta$'),
-
-    ('uavcan', r'sensors/.*\.cpp$', None, r'^_orb_topic$'),
-    ]
-        special_cases_pub = [(a, re.compile(b), re.compile(c) if c is not None else None, re.compile(d))
-                                   for a,b,c,d in special_cases_pub]
-        self._publications = PubSub(True, topic_blacklist,
-                [r"\borb_advertise(_multi|_queue|_multi_queue|)\b\(([^,)]+)",
-                 r"\borb_publish_auto()\b\(([^,)]+)"],
-                special_cases_pub)
-
-
-    def _get_current_module(self):
-        if len(self._current_module) == 0:
+        # note: the source-file-string is pre-processed to remove whitespace -- regexes should ignore whitespace
+        # note:  the regexes should have at least 3 capture groups '()'; otherwise they break downstream code
+        capture_cases_sub = [r"\borb_subscribe(?:_multi|)\b\s*\(\s*(ORB_ID)\s*\(\s*(\w+)",
+                             r"(?:uORB::)Subscription(?:Interval|)\s+\w+\s*[\{\(]\s*(ORB_ID)\s*\(\s*(\w+)",
+                             r"(?:uORB::)Subscription(?:Data|MultiArray|Blocking|)\s*(<)\s*(\w+)",
+                             r"(?:uORB::)SubscriptionCallbackWorkItem\s+\w+\s*\{\s*this,\s*(ORB_ID)\((\w+)",
+                            ]
+        self._subscriptions = Subscriptions( self._topic_blacklist, capture_cases_sub)
+
+        # note: the source-file-string is pre-processed to remove whitespace -- regexes should ignore whitespace
+        # note: the regexes should have at least 3 capture groups '()'; otherwise they break downstream code
+        capture_cases_pub = [r"(?:uORB::)Publication(?:Data|Multi|)\s*(<)(\w+)>",
```

Comment:
Those two lines capture different cases.  For instance, take the source line: 
```
uORB::Publication<vehicle_attitude_setpoint_s>	_fw_virtual_att_sp_pub{ORB_ID(fw_virtual_attitude_setpoint)};
```
the first regex (L262) captures the template-argument-topic, and the fourth regex (L265) captures the ORB_ID-argument topic.

It would arguably be more efficient to capture both topics in one regex, but this simplifies the handle code in the match function.